### DISCUSSION TITLE
Tags System

### DISCUSSION THREAD

**Discussion by snowy-jaguar** [Zavy86/WikiDocs#193] [Ideas]
I really like the look and feel of WikiDocs for my homelab but the one pain point I have is the lack of a tags system; for example I have the following sections:
- hardware
- software
- services

"Server A" would likely make an appearance in all 3 sections:
- the hardware components it has,
- the software it runs and;
- the services it provides.

At the moment, there's no way I can easily tie all these different places together, in an intuitive way at least. I could manually write tags in the top or bottom of the content files but I see that as error prone.

My idea, since the directory's already map out the content structure, is to add a `tags.[insert filetype]` or `options.[insert filetype]` to the directory's. These could be edited via the same WYSIWYG editor, maybe in a submenu or side panel, and be invisible to the user.

My reason for `options.[insert filetype]` is that you could use this to declare other customisations for a page, such as:
- a different slug
- a different colour
- the pages tags
- a option to hide the page from the nav bar (making it only accessible via a link)

My initial thought for the filetype was `.yml` but i guess `.env`, `.toml` or really anything else could work. These properties could be declared in a high level file in the root directory but I think shorter documents in each content directory better fits with the current file structure.

Edit 01: Just to be clear this discussion page is focusing on the tag system idea, the other usecases of an `options` file are merely examples I thought could justify it.

Edit 02: Misc formatting and spelling fixes

Edit 03: Sentence structure edits to make the paragraphs more readable.

**Comment by Zavy86** [comment]
very interesting, we have to brainstorm a little to find the best implementation way..

**Reply by snowy-jaguar** [reply]
Thanks for considering the idea! 😄
I know the implementations I proposed may not be the best but I'd welcome feedback once you've had a chance to brainstorm.

**Reply by snowy-jaguar** [reply]
@Zavy86 I don't mean to be pushy snd I know development takes time but; I was wondering if you (or maybe one of the other contributers have had a chance to brainstorm a little on a implementation method for this?
I've seen an "issue" for creating shirtger slugs so maybe this could tie in to that in some way

**Reply by leomoon** [reply]
So the main focus of wikidocs is portability and not having a DB.

Not having a DB introduces some challenges like adding a tag feature.

I'll see how we would do this.

**Reply by Zavy86** [reply]
maybe like the sitemap we can parse all documents searching for #tags, build an array and store it into a JSON file..
but if the wiki has many files we could get some performance issues!
or parse only the current file when saving and append data to the JSON tags index file?

possible file structure:

[
  {
    "doc": "homepage"
    "tags": [
      "home",
      "wiki"
    ]
  },
  {
    "doc": "path/of/doc"
    "tags": [
      "test"
    ]
  }
]

or simply use the search function to search hashtags without indexing?

**Reply by snowy-jaguar** [reply]
What were your thoughts on the implementation ideas I posted in my initial post?

Essentially having a dedicated `tags.yml` (or some other filetype, like toml or env, for example) where the tags get their own file and then you parse them in a similar way to the markdown files.

Or having one `tags.toml` in the root directory and then:
- listing all of the tags as `[section]` headings and all of the pages that tag applies to as entities of that heading
- listing all of the pages/filepaths as `[section]` headings and all of the tags that apply to that page as entities of that heading.

I think either of those, if doable, also allows for further customisation down the road.

**Reply by snowy-jaguar** [reply]
Following this up as I'm curious if there's been any movement or decision made on it?

**Comment by ffiesta** [comment]
Hi to all,

After read, I have some concerns!

1. If tags stay in content.md file search option will find it, and it will be messy, and I think its harder somehow hard setup Parsedown files to exclude this
1.1 but if it is decided insert directly on content.md it need to be in beginning, because i think is more easily configure Parsedown system.
1.2 if tags is directly written in content.md search will find it, and it will be big messy, and duplicated values on search
2. alternatively create a automatically [w-tags] function that wi list all tags in line 2, that work like [toc] system but we won't need to write on content.md file. This automatically index all words that we identify as tags: as example i write word **world** and add [[world]].
3. Generate 1 big file or multiple files
4.1 we can generate a big tag file, like sitemap but it will freeze somehow wikidocs when it create this file! This happened now in my wiki with 1098 content.md files with ... xxx files inside!
4.2 we can generate 1 file by content.md file avery time we save, and create index.tag.xml or json file that will index all tags file from folders.
5. We need to add a check on search bar to search only on index tag file, and preventing index all words of all files
6. Alternatively we add check box for search only tags and this tags are written directly on doc identified in someway, like I make now in beginning file i add [unifi] and when i went only this tags in search i write [unifi and it limits the search!

Best regards
